### PR TITLE
[WIP] Support Hiddenfields

### PR DESCRIPTION
**Pull request by ayanai1**
#1308 Hide and unhide Fields  

This Feature allows the user to hide and unhide selected fields.

- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running  JabRef
- [x] Screenshot of feature

### CODE REVIEW

File: src/main/java/net/sf/jabref/model/entry/BibEntry.java [hunk 1]
```diff
@@ -1,669 +1,683 @@
...
-                return Collections.emptySet();
-            } else {
-                HashSet<String> words = new HashSet<>(EntryUtil.getStringAsWords(fieldValue));
-                fieldsAsWords.put(fieldName, words);
-                return words;
-            }
-        }
-    }
-}
+/*  Copyright (C) 2003-2015 JabRef contributors.
```
Comment (koppor) [inline src/main/java/net/sf/jabref/model/entry/BibEntry.java]: You seem to have line ending issues. Please fix your commit.

**Comment by Siedlerchr** [comment]
The line ending issue is probably a git config option:
http://stackoverflow.com/questions/10418975/how-to-change-line-ending-settings
On windows it should be set to autorcrlf = true

File: src/main/java/net/sf/jabref/gui/actions/HideAction.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+
+        String fieldName = fieldEditor.getFieldName();
+        if (bibEntry.isPresent()) {
+            BibEntry entry = bibEntry.get();
+            if (fieldEditor.isHidden()) {
+                System.out.println("unhiding: " + fieldName);
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/actions/HideAction.java]: Please always use LOGGER. Read about https://logging.apache.org/log4j/2.x/manual/configuration.html. Have a look at `src/main/resources/log4j2.xml`.

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 3]
```diff
@@ -1380,7 +1381,7 @@ public void actionPerformed(ActionEvent e) {
                     LabelPatternPreferences.fromPreferences(Globals.prefs));
 
             // Store undo information:
-            panel.getUndoManager().addEdit(new UndoableKeyChange(panel.getDatabase(), entry, oldValue, entry.getCiteKey()));
+            panel.getUndoManager().addEdit(new UndoableKeyChange(panel.getDatabase(), entry, (String) oldValue, entry.getCiteKey()));
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: Do not cast

**Comment by koppor** [comment]
If that code does not work, the reason is a wrong design decision.

1. Plesae create HiddenFieldEditor for a hidden field. It treats fileds prefixed with `_`
2. The EntryEditorTab contains both HiddenFieldEditor and FieldEditor. One is shown, one is hidden.
3. "Hide" hides FieldEditor and shows HiddenFieldEditor
4. "Show" hides HiddenFieldEditor, shows FieldEditor

 The show/replace above just hides/shows the

File: src/main/java/net/sf/jabref/gui/PreambleEditor.java [hunk 4]
```diff
@@ -93,7 +93,7 @@ protected boolean accept(Component c) {
 
         String content = base.getPreamble();
 
-        ed = new TextArea(Localization.lang("Preamble"), content == null ? "" : content);
+        ed = new TextArea(Localization.lang("Preamble"), content == null ? "" : content, null);
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/PreambleEditor.java]: Never pass "null". We always use "Optional" for that. And we always use "Objects.requireNonNull" for the Paramter check.

**Comment by koppor** [comment]
Suspended until September, 5th (if no one else is going to discuss https://github.com/JabRef/jabref/pull/1655#issuecomment-239390322 with @ayanai1).

**Comment by ayanai1** [comment]
Hide and Show field is now working just fine for the required fields(figure 1) and the optional fields(figure 2)

hiddenreq
Figure 1. required fields

hiddenopt
Figure 2. optional fields

**Comment by ayanai1** [comment]
Hide and Show field for required(figure 1) and optional(figure2) fields is working:

hiddenreq
Figure 1: required fields

hiddenopt
Figure 2: optional Fields

The Bibtex scource tab(figure 3) shows which fields are hidden too:
bibsource
Figure 3: Bibtex scource tab

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 5]
```diff
@@ -305,25 +305,47 @@ private void setupFieldPanels() {
...
         addSourceTab();
     }
 
-    private void addGeneralTabs() {
+    private void addAbstractReviewTabs() {
         EntryEditorTabList tabList = Globals.prefs.getEntryEditorTabList();
-        for (int i = 0; i < tabList.getTabCount(); i++) {
-            EntryEditorTab newTab = new EntryEditorTab(frame, panel, tabList.getTabFields(i), this, false,
-                    false, tabList.getTabName(i));
+        for (int i = 1; i < tabList.getTabCount(); i++) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: Sure it is not zero based? Normally you would do i=0 and <= Count

Comment (ayanai1) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: yes zero ist the GeneralTab which has its own Method now

Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: Okay, then it would be nice if you could add a simple comment above to make it easier for others...
Jusdt like: //starting at 1 because general tab is zero and has its own method 
or whatever

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 6]
```diff
@@ -288,32 +289,70 @@ private void setupFieldPanels() {
...
 
-    private void addGeneralTabs() {
+    private void addAbstractReviewTabs() {
         EntryEditorTabList tabList = Globals.prefs.getEntryEditorTabList();
-        for (int i = 0; i < tabList.getTabCount(); i++) {
-            EntryEditorTab newTab = new EntryEditorTab(frame, panel, tabList.getTabFields(i), this, false,
-                    false, tabList.getTabName(i));
+        //i starts with 1 because 0 is the GeneralTab which has its own method
+        for (int i = 1; i < tabList.getTabCount(); i++) {
+            List<String> tabFields = new ArrayList<>();
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: I think you could extract the for loop below into a new method ``public static ArrayList<String > addFieldAndHiddenField(ArrayList<String> list, String fieldName) {...}`` since you do this 3 times in this class.

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 7]
```diff
@@ -288,32 +289,70 @@ private void setupFieldPanels() {
...
+            }
+            EntryEditorTab newTab = new EntryEditorTab(frame, panel, tabFields, this, true, false,
+                    tabList.getTabName(i));
             if (newTab.fileListEditor != null) {
                 fileListEditor = newTab.fileListEditor;
             }
             tabbed.addTab(tabList.getTabName(i), newTab.getPane());
             tabs.add(newTab);
         }
+
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: please delete this empty line.

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 8]
```diff
@@ -338,8 +377,16 @@ private void addOtherTab(List<String> otherFields) {
 
     private List<String> addRequiredTab(EntryType type) {
         List<String> requiredFields = type.getRequiredFieldsFlat();
+        // build a list of all required fields, where each required field is followed by the hidden counter part
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: And if you extract the for loop, you could write something like that in the javadoc

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java [hunk 9]
```diff
@@ -151,8 +156,19 @@ private void setupPanel(JabRefFrame frame, BasePanel bPanel, boolean addKeyField
                         false);
                 defaultHeight = 0;
             } else {
-                fieldEditor = new TextArea(field, null);
-                bPanel.frame().getGlobalSearchBar().getSearchQueryHighlightObservable().addSearchListener((TextArea) fieldEditor);
+                if (field.startsWith("_")) {
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java]: Please extract to new Method ``isHiddenFIeld(field)``

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java [hunk 10]
```diff
@@ -285,9 +301,37 @@ public void setEntry(BibEntry entry) {
...
             for (FieldEditor editor : editors.values()) {
-                String toSet = entry.getField(editor.getFieldName()).orElse("");
-                if (!toSet.equals(editor.getText())) {
-                    editor.setText(toSet);
+                String fieldName = editor.getFieldName();
+
+                boolean displayableFieldValueExists;
+                boolean hiddenFieldValueExists;
+                boolean currentFieldNameIsHidden = fieldName.startsWith("_");
+                if (currentFieldNameIsHidden) {
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java]: then use here again ``isHiddenField(field)``

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java [hunk 11]
```diff
@@ -285,9 +301,37 @@ public void setEntry(BibEntry entry) {
...
+                    displayableFieldValueExists = !entry.getField(fieldName.substring(1)).orElse("").isEmpty();
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java]: why don't just replace fieldName.substring(1) with "_"?

Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java]: If I think of it you could also extract ``!entry.getField(fieldName.substring(1)).orElse("").isEmpty()`` to a new method ``entryHasFieldName(entry, fieldName)``

File: src/main/java/net/sf/jabref/gui/fieldeditors/contextmenu/FieldTextMenu.java [hunk 12]
```diff
@@ -106,26 +111,34 @@ private void maybeShowPopup(MouseEvent e) {
...
+                inputMenu.add(new CaseChangeMenu((JTextComponent) field.getTextComponent()));
+                inputMenu.add(new ConversionMenu((JTextComponent) field.getTextComponent()));
+                inputMenu.addSeparator();
+                protectedTermsMenu = new ProtectedTermsMenu((JTextComponent) field.getTextComponent());
+                inputMenu.add(protectedTermsMenu);
+            }
         }
     }
 
+
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/fieldeditors/contextmenu/FieldTextMenu.java]: delete empty line.

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java [hunk 13]
```diff
@@ -285,9 +301,37 @@ public void setEntry(BibEntry entry) {
...
+                    hiddenFieldValueExists = !entry.getField(fieldName).orElse("").isEmpty();
+                } else {
+                    displayableFieldValueExists = !entry.getField(fieldName).orElse("").isEmpty();
+                    hiddenFieldValueExists = !entry.getField("_" + fieldName).orElse("").isEmpty();
+                }
+
+                boolean currentFieldEditorIsVisibleFieldEditor = !currentFieldNameIsHidden;
+                boolean isVisible;
+                if (currentFieldEditorIsVisibleFieldEditor) {
+                    // do not display if hidden field value exists (because other editor is shown)
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditorTab.java]: Please stat comments with capital letters

File: src/main/resources/l10n/JabRef_in.properties [hunk 14]
```diff
@@ -2295,8 +2295,15 @@ Select_first_entry=
...
+Hide_field=
+Optionals_fields=
+Show_Field=
 should_be_an_integer_or_normalized=
 should_be_normalized=
 Copy_BibTeX_key_and_link=
 empty_BibTeX_key=
 BibLaTeX_field_only=
+Driver_error=
+The_BibEntry_you_currently_work_on_has_been_deleted_on_the_shared_side._Hit_"Keep"_to_recover_the_entry=
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_in.properties]: This is not in the language files of this PR. Please fix translations. Proposal: Use the translation files of the master branch and readd requried translations.

File: src/main/resources/l10n/JabRef_in.properties [hunk 15]
```diff
@@ -2295,8 +2295,15 @@ Select_first_entry=
 Select_last_entry=
 
 Invalid_ISBN\:_'%0'.=
+
+Hide_field=
+Optionals_fields=
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_in.properties]: Why the `s` at the end of `Optionals`? - Why is this included in this ?R

File: src/main/resources/l10n/JabRef_it.properties [hunk 16]
```diff
@@ -2158,6 +2158,146 @@ Cleared_'%0'_for_%1_entries=Reinizializzati_'%0'_per_%1_voce/i
...
+Click_here_to_learn_about_the_migration_of_pre-3.6_databases.=
+Opens_JabRef's_Facebook_page=
+Opens_JabRef's_blog=
+Opens_JabRef's_website=
+Opens_a_link_where_the_current_development_version_can_be_downloaded=
+See_what_has_been_changed_in_the_JabRef_versions=
+Referenced_BibTeX_key_does_not_exist=
+Finished_downloading_full_text_document_for_entry_%0.=
+Full_text_document_download_failed_for_entry_%0.=
+Look_up_full_text_documents=
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_it.properties]: Not part of your PR. See above.

File: src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
+package net.sf.jabref.gui.fieldeditors;
+
+import java.util.Objects;
+
+/**
+ * Indicates that the field of this TextArea can be hidden
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java]: why hidden? I think it is for visible fields?

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 18]
```diff
@@ -361,6 +399,18 @@ private void addOptionalTab(EntryType type) {
         tabs.add(optionalPanel);
     }
 
+    public static List<String> addFieldAndHiddenField(List<String> list) {
+        List<String> returnList = new ArrayList<>();
+        /**
+         * build a list of all required fields, where each required field is followed by the hidden counter part
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: please move the javadoc above the method

Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: please move the javadoc above the method

File: src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java (hunk 17 above)
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java]: Please address the comment. Whether write why you would leave this as it is or change it, but please dont let it unanswered.

File: src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java [hunk 19]
```diff
@@ -286,28 +287,34 @@ private void setupFieldPanels() {
...
+                .filter(f -> !displayedFields.contains(f)).collect(Collectors.toList());
         otherFields.remove(BibEntry.KEY_FIELD);
         otherFields.removeAll(Globals.prefs.getCustomTabFieldNames());
 
         if (!otherFields.isEmpty()) {
             addOtherTab(otherFields);
         }
 
+        addGeneralTab();
         // general fields from preferences
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/entryeditor/EntryEditor.java]: Shouldn't this comment be one line above?

File: src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java (hunk 17 above)
Comment (ayanai1) [inline src/main/java/net/sf/jabref/gui/fieldeditors/TextAreaForVisibleField.java]: The fields in that TextArea are visible and can be hidden if wanted

**Comment by tschechlovdev** [comment]
Please fix travis, otherwise it's ok from my side 👍

File: CHANGELOG.md [hunk 20]
```diff
@@ -58,7 +58,8 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Generalize German translation of database into "Datenbank"
 - Improve language quality of the German translation of shared database
 - Change "Recent files" to "Recent databases" to keep the file menu consistent
-- Customized importer files need to be slightly changed since the class `ImportFormat` was renamed to `Importer`
```
Comment (koppor) [inline CHANGELOG.md]: Why is this change deleted?

**Comment by koppor** [comment]
Please comment on the layout issues. I think, they cannot be solved, can they?

File: CHANGELOG.md (hunk 20 above)
Comment (ayanai1) [inline CHANGELOG.md]: this happend because of a conflict error and is a mistake i put back right away

**Comment by ayanai1** [comment]
@koppor the problem here is that we do not have any time left and the layout is beeing created by a DefaultFormBuilder if we had more time to design the layout ourselfs this issue could be resolved but unfortunatly we don't.

**Comment by Siedlerchr** [comment]
The DefaultFormBuilder stuff can be manually adjusted, but it's a complicated and frustrating mess... :fearful:

**Comment by koppor** [comment]
OK, thank you for the information. Then I treat this as nice try and a learnied experience with coding UIs.
